services/scan/src/cvrs/export.ts

```diff
@@ -0,0 +1,184 @@
+      !skipImages;
+
+    const cvr = buildCastVoteRecord(
+      id,
+      batchId,
+      batchLabel || '',
+      (interpretation[0].type === 'InterpretedBmdPage' &&
+        interpretation[0].ballotId) ||
+        (interpretation[1].type === 'InterpretedBmdPage' &&
+          interpretation[1].ballotId) ||
+        unsafeParse(BallotIdSchema, id),
+      electionDefinition.election,
+      [
+        {
+          interpretation: interpretation[0],
+          contestIds: isHmpbPage(interpretation[0])
+            ? store.getContestIdsForMetadata(
+                interpretation[0].metadata,
+                electionDefinition
+              )
+            : undefined,
+        },
+        {
+          interpretation: interpretation[1],
+          contestIds: isHmpbPage(interpretation[1])
+            ? store.getContestIdsForMetadata(
+                interpretation[1].metadata,
+                electionDefinition
+              )
+            : undefined,
+        },
+      ],
+      includeImages && isHmpbSheet(interpretation)
+        ? mapSheet(
+            interpretation,
+            ({ metadata }) =>
+              store.getBallotPageLayoutForMetadata(
+                metadata,
+                electionDefinition
+              ) as BallotPageLayout
+          )
+        : undefined
+    );
+
+    if (cvr) {
+      let cvrMaybeWithBallotImages = cvr;
+
+      // if write-in adjudication & there are write-ins in this CVR, we augment record with ballot images
+      if (includeImages && isHmpbSheet(interpretation)) {
+        const [frontHasWriteIns, backHasWriteIns] = cvrHasWriteIns(
+          electionDefinition.election,
+          cvr
+        );
+        if (frontHasWriteIns) {
+          const frontFilenames = store.getBallotFilenames(id, 'front');
+          if (frontFilenames) {
+            frontImage.normalized = await loadImagePathShrinkBase64(
+              frontFilenames.normalized,
+              CvrBallotImageScale
+            );
+          }
+        }
+
+        if (backHasWriteIns) {
+          const backFilenames = store.getBallotFilenames(id, 'back');
+          if (backFilenames) {
+            backImage.normalized = await loadImagePathShrinkBase64(
+              backFilenames.normalized,
+              CvrBallotImageScale
+            );
+          }
+
+          cvrMaybeWithBallotImages = addBallotImagesToCvr(cvr, [
+            frontImage,
+            backImage,
+          ]);
+        }
+      }
+
+      yield cvrMaybeWithBallotImages;
+    }
+  }
+}
+
+/**
+ * Export all CVRs newline-delimited JSON.
+ */
+export function exportCastVoteRecordsAsNdJson({
+  store,
+  skipImages,
+  orderBySheetId,
+}: {
+  store: Store;
+  skipImages?: boolean;
+  orderBySheetId?: boolean;
+}): NodeJS.ReadableStream {
+  return Readable.from(
+    mapAsync(
+      exportCastVoteRecords({ store, skipImages, orderBySheetId }),
+      (cvr) => `${JSON.stringify(cvr)}\n`
+    )
```

Comment:
My understanding of generating a [Readable from an iterator](https://nodejs.org/api/stream.html#streamreadablefromiterable-options) is that this will will mean items will become available in the stream as they are generated, as opposed to a Readable from an `Buffer` or `string`, correct?

Is there anything that we need to do to indicate that the Readable is closed and will no longer emit data once we've worked through all the sheets? Or is that being handled somewhere already?

Reply:
Looking at the [implementation of `Readable.from`](https://github.com/nodejs/node/blob/badc631d778bd13043770a34091ead49291a73d9/lib/internal/streams/from.js#L88-L89), you can see that when the iterator is done it pushes `null` to the stream, which should end it properly. For a generator function `next` will automatically return `{ done: true }` when the function exits, so this should be sufficient without any explicit cleanup.